crates/oxc_linter/src/rules/eslint/no_control_regex.rs

```diff
@@ -0,0 +1,351 @@
+                    }
+                    // even # of slashes
+                    if num_slashes % 2 == 0 {
+                        continue;
+                    }
+                }
+
+                if ctl.starts_with(r"\x") || ctl.starts_with(r"\u") {
+                    let mut numeric_part = &ctl[2..];
+
+                    // extract numeric part from \u{00}
+                    if numeric_part.starts_with('{') {
+                        let has_unicode_flag = match flags {
+                            Some(flags) if flags.contains(RegExpFlags::U) => true,
+                            _ => {
+                                continue;
+                            }
+                        };
+
+                        // 1. Unicode control pattern is missing a curly brace
+                        //    and is therefore invalid. (note: we may want to
+                        //    report this?)
+                        // 2. Unicode flag is missing, which is needed for
+                        //    interpreting \u{`nn`} as a unicode character
+                        if !has_unicode_flag || !numeric_part.ends_with('}') {
+                            continue;
+                        }
+
+                        numeric_part = &numeric_part[1..numeric_part.len() - 1];
+                    }
+
+                    match u32::from_str_radix(numeric_part, 16) {
+                        Err(_) => continue,
+                        Ok(as_num) if as_num > 0x1f => continue,
+                        Ok(_) => { /* noop */ }
+                    }
+                }
+
+                violations.push(ctl);
+            }
+
+            if !violations.is_empty() {
+                let violations = violations.join(", ");
+                context.diagnostic(NoControlRegexDiagnostic(violations.into(), span));
+            }
+        }
+    }
+}
+
+fn extract_flags<'a>(args: &'a oxc_allocator::Vec<'a, Argument<'a>>) -> Option<RegExpFlags> {
+    if args.len() > 1 &&
+    let Argument::Expression(Expression::StringLiteral(flag_arg)) = &args[1] {
+        // TODO: how should we handle invalid flags?
+        let mut flags = RegExpFlags::empty();
+        for ch in flag_arg.value.chars() {
+            let flag = RegExpFlags::try_from(ch).ok()?;
+            // TODO: should we check for duplicates?
+            flags |= flag;
+        }
+        Some(flags)
+    } else {
+        None
+    }
+}
+
+/// Returns the regex pattern inside a node, if it's applicable.
+///
+/// e.g.:
+/// * /foo/ -> "foo"
+/// * new RegExp("foo") -> foo
+///
+/// note: [`RegExpFlags`] and [`Span`]s are both tiny and cloneable.
+fn regex_pattern<'a>(node: &AstNode<'a>) -> Option<(&'a Atom, Option<RegExpFlags>, Span)> {
+    let kind = node.kind();
+    match kind {
+        // regex literal
+        AstKind::RegExpLiteral(reg) => Some((&reg.regex.pattern, Some(reg.regex.flags), reg.span)),
+
+        // FIXME: we need a more graceful way to handle NewExpr/CallExprs
```

Comment:
What are the problems are you facing?

Reply:
Look at how much of the code is duplicated between both branches.  It may be nice to have a `CallLikeExpression` or something similar